src/compas/datastructures/tree/tree.py

```diff
@@ -0,0 +1,259 @@
+        The name of the datastructure.
+    attributes : dict[str, Any]
+        User-defined attributes of the datastructure.
+    root: :class:`~compas.datastructures.TreeNode`
+        The root node of the tree.
+    nodes: set[:class:`~compas.datastructures.TreeNode`]
+        The nodes of the tree.
+    leaves: generator[:class:`~compas.datastructures.TreeNode`]
+        A generator of the leaves of the tree.
+
+    Examples
+    --------
+    >>> from compas.datastructures import Tree, TreeNode
+    >>> tree = Tree()
+    >>> root = TreeNode('root')
+    >>> branch = TreeNode('branch')
+    >>> leaf1 = TreeNode('leaf1')
+    >>> leaf2 = TreeNode('leaf2')
+    >>> tree.add_root(root)
+    >>> root.add(branch)
+    >>> branch.add(leaf1)
+    >>> branch.add(leaf2)
+    >>> print(tree)
+    <Tree with 4 nodes, 1 branches, and 2 leaves>
+    >>> tree.print()
+    <TreeNode root>
+        <TreeNode branch>
+            <TreeNode leaf2>
+            <TreeNode leaf1>
+
+    """
+
+    JSONSCHEMA = {}
+
+    def __init__(self, name: str = None, attributes: dict = None):
+        super(Tree, self).__init__()
+        self.name = name
+        self.attributes = attributes or {}
+        self._root = None
+        self._nodes = set()
+
+    @property
+    def data(self):
+        def get_node_data(node):
+            return {
+                "name": node.name,
+                "attributes": node.attributes,
+                "children": [get_node_data(child) for child in node.children],
+            }
+
+        return {
+            "name": self.name,
+            "root": get_node_data(self.root),
+            "attributes": self.attributes,
+        }
+
+    @data.setter
+    def data(self, data):
+        self.name = data["name"]
+        self.attributes = data["attributes"]
+
+        def node_from_data(data):
+            node = TreeNode(data["name"], data["attributes"])
+            for child in data["children"]:
+                node.add(node_from_data(child))
+            return node
+
+        self.add_root(node_from_data(data["root"]))
+
+    @property
+    def root(self):
+        return self._root
+
+    def add_root(self, node: TreeNode):
+        """Add a root node to the tree."""
+        if not isinstance(node, TreeNode):
+            raise TypeError("The node is not a TreeNode object.")
+        if not node.is_root:
+            raise ValueError("The node is already part of another tree.")
+        self._root = node
+        node._tree = self
+        self._nodes.add(node)
+
+    def add(self, node: TreeNode, parent: TreeNode):
+        """Add a node to the tree."""
+        if not isinstance(node, TreeNode):
+            raise TypeError("The node is not a TreeNode object.")
+        if self.root is None:
+            raise ValueError("The tree has no root node, use add_root() first.")
+        else:
+            parent.add(node)
+
+    @property
+    def nodes(self):
+        return self._nodes
+
+    def remove_root(self):
```

Comment:
I would make this method private (underscored), otherwise the API surface is confusing.

Reply:
makes sense